File: app/src/main/java/lt/zilinskas/marius/testio/adapters/ServersAdapter.java
package lt.zilinskas.marius.testio.adapters;


import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import lombok.Getter;
import lombok.Setter;
import lt.zilinskas.marius.testio.R;
import lt.zilinskas.marius.testio.adapters.ServersAdapter.CustomViewHolder;
import lt.zilinskas.marius.testio.api.entities.Server;

public class ServersAdapter extends RecyclerView.Adapter<CustomViewHolder> {
    @Setter
    private List<Server> servers;
    private CompositeDisposable compositeDisposable;

    public ServersAdapter(List<Server> servers) {
        this.servers = servers;
    }

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.list_row_server, viewGroup, false);
        return new CustomViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder customViewHolder, int i) {
        Server server = servers.get(i);
        customViewHolder.getName().setText(server.getName());
        customViewHolder.getDistance().setText(String.valueOf(server.getDistance()));
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        compositeDisposable = new CompositeDisposable();
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        if (compositeDisposable != null && !compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
        super.onDetachedFromRecyclerView(recyclerView);
    }

    @Override
    public int getItemCount() {
        return servers.size();
    }


    @Getter
    class CustomViewHolder extends RecyclerView.ViewHolder {
        private View parent;
        private TextView name;
        private TextView distance;

        CustomViewHolder(@NonNull View itemView) {
            super(itemView);
            this.parent = itemView;
            initViews();
        }

        private void initViews() {
            name = parent.findViewById(R.id.nameTextView);
            distance = parent.findViewById(R.id.distanceTextView);
        }
    }
}

File: app/src/main/java/com/alex/tesonettesttask/ui/adapters/ServersAdapter.kt
package com.alex.tesonettesttask.ui.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.alex.tesonettesttask.R
import com.alex.tesonettesttask.logic.entities.Server
import kotlinx.android.synthetic.main.server_item.view.*

class ServersAdapter(private val servers: List<Server>) : RecyclerView.Adapter<ServersAdapter.ServerViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ServerViewHolder {
        return ServerViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.server_item, parent, false))
    }

    override fun onBindViewHolder(viewHolder: ServerViewHolder, position: Int) {
        val serverItem = servers[position]
        with(viewHolder.itemView) {
            server.text = serverItem.name
            distance.text = String.format("%s km", serverItem.distance)
            divider.visibility = if (position == servers.size - 1) View.GONE else View.VISIBLE
        }
    }

    override fun getItemCount() = servers.size

    inner class ServerViewHolder(val view: View) : RecyclerView.ViewHolder(view)
}